aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1359,127 +1356,110 @@ Tensor orgqr(const Tensor& input, const Tensor& tau) {
 
-  // Using 'int' instead of int32_t or int64_t is consistent with the current LAPACK interface
-  // It really should be changed in the future to something like lapack_int that depends on the specific LAPACK library that is linked
-  // or switch to supporting only 64-bit indexing by default.
-  int info;
-  int lwork = -1;
-  int lrwork = -1;
-  int liwork = -1;
-  scalar_t work_query;
-  value_t rwork_query;
-  int iwork_query;
+  TORCH_INTERNAL_ASSERT(input.device() == vectors.device());
+  TORCH_INTERNAL_ASSERT(input.device() == values.device());
 
-  // Run lapackSyevd once, first to get the optimum work size.
-  // Since we deal with batches of matrices with the same dimensions, doing this outside
-  // the main loop saves (batch_size - 1) workspace queries which would provide the same result
-  // and (batch_size - 1) calls to allocate and deallocate workspace using at::empty()
-  lapackSyevd<scalar_t, value_t>(jobz, uplo, n, v_data, lda, w_data, &work_query, lwork, &rwork_query, lrwork, &iwork_query, liwork, &info);
+  // eigenvalues are always real-valued
+  ScalarType real_dtype = toValueType(input.scalar_type());
+  TORCH_INTERNAL_ASSERT(values.scalar_type() == real_dtype);
+  TORCH_INTERNAL_ASSERT(input.scalar_type() == vectors.scalar_type());
 
-  lwork = std::max<int>(1, real_impl<scalar_t, value_t>(work_query));
-  Tensor work = at::empty({lwork}, v.options());
-  liwork = std::max<int>(1, iwork_query);
-  Tensor iwork = at::empty({liwork}, at::kInt);
+  TORCH_INTERNAL_ASSERT(infos.scalar_type() == at::kInt);
+  TORCH_INTERNAL_ASSERT(infos.device() == input.device());
+  TORCH_INTERNAL_ASSERT(infos.numel() == std::max<int64_t>(1, batchCount(input)));
+  TORCH_INTERNAL_ASSERT(infos.is_contiguous());
 
-  Tensor rwork;
-  value_t* rwork_data = nullptr;
-  if (isComplexType(at::typeMetaToScalarType(v.dtype()))) {
-    lrwork = std::max<int>(1, rwork_query);
-    rwork = at::empty({lrwork}, w.options());
-    rwork_data = rwork.data_ptr<value_t>();
+  // if 'vectors' has no elements we can modify it
+  if (vectors.numel() == 0) {
+    vectors.resize_(input.sizes(), MemoryFormat::Contiguous);
+    vectors.transpose_(-2, -1);  // make 'vectors' to have Fortran contiguous memory layout
   }
 
-  // Now call lapackSyevd for each matrix in the batched input
-  for (const auto i : c10::irange(batch_size)) {
-    scalar_t* v_working_ptr = &v_data[i * v_matrix_stride];
-    value_t* w_working_ptr = &w_data[i * w_stride];
-    lapackSyevd<scalar_t, value_t>(jobz, uplo, n, v_working_ptr, lda, w_working_ptr, work.data_ptr<scalar_t>(), lwork, rwork_data, lrwork, iwork.data_ptr<int>(), liwork, &info);
-    infos[i] = info;
-    // The current behaviour for Linear Algebra functions to raise an error if something goes wrong or input doesn't satisfy some requirement
-    // therefore return early since further computations will be wasted anyway
-    if (info != 0) {
-      return;
-    }
+  // if 'values' has no elements we can modify it
```

Comment:
Same with another recent PR, would you elaborate on the juxtaposition of this code block with the internal asserts for values' shape later?

Reply:
This function is intended to work only for correctly sized outputs in column-major contiguous memory format or with zero number of elements that is resized.
The out= variant here currently is implemented via a temporary tensor for the result and copy. out= variant that fuses the copy when out tensors are lapack-compatible will be implemented in other PR.